### PR TITLE
Return the matching indices from `has_keys()`

### LINKED ISSUE #7285: keystore.has_keys behaviour should return the indicies found
Follow up to the comment here: https://github.com/paritytech/polkadot/pull/1740#discussion_r502037238

The behaviour of the `has_keys` would be slow in cases such as:

https://github.com/paritytech/polkadot/blob/9ce9eccae4faa7935bbca9a4200e0978168a46f9/node/network/availability-distribution/src/lib.rs#L582-L586

where the list of keys is iterated so that the first key found is returned. However, in case of a remote signer, this loop would be quite slow.

`has_keys` behaviour should be modified so that the index of the first element found is returned so that this index can be used to return the public key.

### PR DESCRIPTION
**Pull request by hirschenberger**
This is a preparation step to land an optimization by avoiding calling that
function multiple times on a remote signer.

fixes #7285

Thank you for your Pull Request!

Before you submitting, please check that:

- [ ] You added a brief description of the PR, e.g.:
  - What does it do?
  - What important points reviewers should know?
  - Is there something left for follow-up PRs?
- [ ] You labeled the PR appropriately if you have permissions to do so:
  - [ ] `A*` for PR status (**one required**)
  - [ ] `B*` for changelog (**one required**)
  - [ ] `C*` for release notes (**exactly one required**)
  - [ ] `D*` for various implications/requirements
  - [ ] Github's project assignment
- [ ] You mentioned a related issue if this PR related to it, e.g. `Fixes #228` or `Related #1337`.
- [ ] You asked any particular reviewers to review. If you aren't sure, start with GH suggestions.
- [ ] Your PR adheres to [the style guide](https://github.com/paritytech/substrate/blob/master/docs/STYLE_GUIDE.md)
  - In particular, mind the maximal line length of 100 (120 in exceptional circumstances).
  - There is no commented code checked in unless necessary.
  - Any panickers have a proof or removed.
- [ ] You bumped the runtime version if there are breaking changes in the **runtime**.
- [ ] You updated any rustdocs which may have changed
- [ ] Has the PR altered the external API or interfaces used by Polkadot? Do you have the corresponding Polkadot PR ready?

Refer to [the contributing guide](https://github.com/paritytech/substrate/blob/master/docs/CONTRIBUTING.adoc) for details.

After you've read this notice feel free to remove it.
Thank you!

✄ -----------------------------------------------------------------------------

### CODE REVIEW

**Comment by hirschenberger** [comment]
Can we run the pipeline again?

File: client/keystore/src/local.rs [hunk 1]
```diff
@@ -277,9 +277,13 @@ impl SyncCryptoStore for LocalKeystore {
...
 
-	fn has_keys(&self, public_keys: &[(Vec<u8>, KeyTypeId)]) -> bool {
-		public_keys.iter()
-			.all(|(p, t)| self.0.read().key_phrase_by_type(&p, *t).ok().flatten().is_some())
+	fn has_keys(&self, public_keys: &[(Vec<u8>, KeyTypeId)]) -> Vec<usize> {
+		public_keys.iter().enumerate().fold(Vec::new(), |mut idxs, (i, (p, t))| {
+			if self.0.read().key_phrase_by_type(&p, *t).ok().flatten().is_some() {
+				idxs.push(i);
+			}
+			idxs
```
Comment (thiolliere) [inline client/keystore/src/local.rs:285]: but this is changing the definition, before it requires all key to exist, now it requires only 1 key to exist.

I'm not very aware of keystore but this seems quite a different behavior to me

**Comment by thiolliere** [comment]
you can merge master to trigger CI

File: client/keystore/src/local.rs (hunk 1 above)
Comment (hirschenberger) [reply client/keystore/src/local.rs:285]: Indeed, but I think that's what @bkchr wanted?

File: primitives/keystore/src/lib.rs [hunk 2]
```diff
@@ -274,8 +275,9 @@ pub trait SyncCryptoStore: CryptoStore + Send + Sync {
 
 	/// Checks if the private keys for the given public key and key type combinations exist.
 	///
-	/// Returns `true` iff all private keys could be found.
-	fn has_keys(&self, public_keys: &[(Vec<u8>, KeyTypeId)]) -> bool;
+	/// Returns the indices of the matching keys in `public_keys`. Call `is_empty()` on the returned
+	/// list to check if no keys match.
+	fn has_keys(&self, public_keys: &[(Vec<u8>, KeyTypeId)]) -> Vec<usize>;
```
Comment (bkchr) [inline primitives/keystore/src/lib.rs:280]: First, sorry for the late reply.

Now after enough time, I actually think we should return some enum here.

```
enum HasKeys {
    None,
    Found(Vec<usize>),
}

impl HasKeys {
    fn found_any(&self) -> bool {
        matches!(self, Self::Found(_))
    }

    fn into_found(self) -> impl Iterator<Item = usize> {
         match self {
             Self::None => Vec::new().into_iter(),
             Self::Found(found) => found.into_iter(),
        }
    }
    }
}
```

Probably the api needs to be improved, but you should get the rough idea.

Comment (hirschenberger) [reply primitives/keystore/src/lib.rs:280]: Looks sleek, maybe we can also add a `found_all` function, if you think that's useful?

Comment (bkchr) [reply primitives/keystore/src/lib.rs:280]: Yeah.

**Comment by hirschenberger** [comment]
Upps, a rebase was no good idea :-(

**Comment by hirschenberger** [comment]
Closing as this PS is messed up. Cleaned up version is #8619

**Pull request closed without merging**